commitrmnocb/observation.go

```diff
@@ -0,0 +1,170 @@
+		observation = Observation{
+			MerkleRoots: p.ObserveMerkleRoots(ctx, previousOutcome.RangesSelectedForReport),
+			GasPrices:   []cciptypes.GasPriceChain{},
+			TokenPrices: []cciptypes.TokenPrice{},
+			FChain:      p.ObserveFChain(),
+		}
+
+	case WaitingForReportTransmission:
+		observation = Observation{
+			OffRampNextSeqNums: p.ObserveOffRampNextSeqNums(ctx),
+			FChain:             p.ObserveFChain(),
+		}
+
+	default:
+		p.lggr.Warnw("Unexpected state", "state", nextState)
+		return observation.Encode()
+	}
+
+	p.lggr.Infow("Observation", "observation", observation)
+	return observation.Encode()
+}
+
+// ObserveOffRampNextSeqNums Observe the next sequence numbers for each source chain from the OffRamp
+func (p *Plugin) ObserveOffRampNextSeqNums(ctx context.Context) []plugintypes.SeqNumChain {
+	supportsDestChain, err := p.supportsDestChain(p.nodeID)
+	if err != nil {
+		p.lggr.Warnw("call to SupportsDestChain failed", "err", err)
+		return nil
+	}
+
+	if supportsDestChain {
+		sourceChains := p.knownSourceChainsSlice()
+		offRampNextSeqNums, err := p.ccipReader.NextSeqNum(ctx, sourceChains)
+		if err != nil {
+			p.lggr.Warnw("call to NextSeqNum failed", "err", err)
+			return nil
+		}
+
+		if len(offRampNextSeqNums) != len(sourceChains) {
+			p.lggr.Warnf("call to NextSeqNum returned unexpected number of seq nums, got %d, expected %d",
+				len(offRampNextSeqNums), len(sourceChains))
+			return nil
+		}
+
+		result := make([]plugintypes.SeqNumChain, len(sourceChains))
+		for i := range sourceChains {
+			result[i] = plugintypes.SeqNumChain{ChainSel: sourceChains[i], SeqNum: offRampNextSeqNums[i]}
+		}
+
+		return result
+	}
+
+	return nil
+}
+
+// ObserveMerkleRoots Compute the merkle roots for the given sequence number ranges
+func (p *Plugin) ObserveMerkleRoots(ctx context.Context, ranges []ChainRange) []cciptypes.MerkleRootChain {
+	roots := make([]cciptypes.MerkleRootChain, 0)
+	supportedChains, err := p.supportedChains(p.nodeID)
+	if err != nil {
+		p.lggr.Warnw("call to supportedChains failed", "err", err)
+		return nil
+	}
+
+	for _, chainRange := range ranges {
+		if supportedChains.Contains(chainRange.ChainSel) {
+			msgs, err := p.ccipReader.MsgsBetweenSeqNums(ctx, chainRange.ChainSel, chainRange.SeqNumRange)
+			if err != nil {
+				p.lggr.Warnw("call to MsgsBetweenSeqNums failed", "err", err)
+			} else {
+				root, err := p.computeMerkleRoot(ctx, msgs)
+				if err != nil {
+					p.lggr.Warnw("call to computeMerkleRoot failed", "err", err)
+				} else {
+					merkleRoot := cciptypes.MerkleRootChain{
+						ChainSel:     chainRange.ChainSel,
+						SeqNumsRange: chainRange.SeqNumRange,
+						MerkleRoot:   root,
+					}
+					roots = append(roots, merkleRoot)
+				}
+			}
+		}
+	}
+
+	return roots
+}
+
+// computeMerkleRoot Compute the merkle root of a list of messages
+func (p *Plugin) computeMerkleRoot(ctx context.Context, msgs []cciptypes.Message) (cciptypes.Bytes32, error) {
+	hashes := make([][32]byte, 0)
```

Comment:
Style: since you're just appending to it prefer `var hashes [][32]byte` see e.g https://github.com/uber-go/guide/blob/master/style.md#nil-is-a-valid-slice